### ISSUE TITLE
Add ability to use inject() for element injector hierarchy

### ISSUE THREAD

**Issue by JeanMeche** [angular/angular#47566]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

When providing a Service in a component, `inject()` triggers `NullInjectorError ` when called within `runInContext` : 

```typescript
@Component({
  selector: 'my-app',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
  providers: [Service],
})
export class AppComponent {
  public constructor(injector: EnvironmentInjector, service: Service) {
    console.log(service); // OK
    console.log(inject(Service)); //OK

    setTimeout(() => {
      injector.runInContext(() => {
        console.log(inject(Service)); //KO
      });
    }, 100);
  }
}
```
This does not happen
* if the service is `providedIn:'root'`
* if the service is provided in the module 

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-ivy-yccqgi?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.component.html

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular 14.2.3
```

**Comment by pkozlowski-opensource** [comment]
This works as designed: `EnvironmentInjector` has only providers configured on the application / NgModule level - it doesn't "see" providers configured on the component level.

Not sure what is your real-life use-case, but you could always do:

```typescript
export class AppComponent {
  public constructor(private _injector: Injector, service: Service) {
    console.log(service); // OK
    console.log(inject(Service)); //OK

    setTimeout(() => {
      console.log(this._injector.get(Service)); //KO
    }, 100);
  }
}
```
https://stackblitz.com/edit/angular-ivy-jujagc?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.component.html

**Comment by JeanMeche** [comment]
Thx @pkozlowski-opensource It this mention somewhere in the documentation ? 

If not, it would be a good idea to add it.

**Comment by pkozlowski-opensource** [comment]
> In my use case, I have objects created outside DI, but they still need to get access to injected services.

Currently your best option is to define those services on the application level.

**Comment by pkozlowski-opensource** [comment]
@alxhub I've marked this is a feature request but please let me know if you've got any additional concerns.

**Comment by antischematic** [comment]
It's possible to create an environment injector that inherits from the node injector, would it be possible to widen the type signature so that `createEnvironmentInjector` allows this without a type cast?

```ts
@Component({
  providers: [Service]
})
class AppComponent {
  constructor(parent: Injector) {
    const injector= createEnvironmentInjector([], parent as EnvironmentInjector)
    setTimeout(() => {
      injector.runInContext(() => {
        console.log(inject(Service)); //KO
      });
    }, 100);
  }
}
```

**Comment by Jonathan002** [comment]
+1 for component injectors. runInContext. I need one for each of my canvas components. I have a canvas game component that would like to `new CharacterSprite(customOptionsPerInstance)`. The `CharacterSprite` is initialized after a server gets data so it is async. Inside `CharacterSprite` I use inject(canvasRef) - which is provided by each canvas component instance. It definitely beats having to do `new CharacterSprite(canvasRef, customOptionsPerInstance)`.

**Comment by akhilsai1969** [comment]
Is this issue still open? If yes, Can you assign it to me?

**Comment by JeanMeche** [comment]
Fixed by #49396, we can now use `runInInjectionContext`
